File: Problems/0300_0399/0341_Flatten_Nested_List_Iterator/Project_Go/solution/operatenestedinteger.go
package solution

import (
	"fmt"
	"strconv"
	"strings"
)

func createListNestedInteger(flds_str string) []*NestedInteger {
	var ni []*NestedInteger
	for len(flds_str) > 0 {
		//	fmt.Printf("flds_str = %s\n", flds_str)
		if flds_str[0] == '[' {
			end_pos, depth := 1, 1
			for depth > 0 && end_pos < len(flds_str) {
				if flds_str[end_pos] == '[' {
					depth++
				} else if flds_str[end_pos] == ']' {
					depth--
				}
				end_pos++
			}
			substr := flds_str[1 : end_pos-1]
			ni = append(ni, createNestedInteger(substr))
			if end_pos < len(flds_str)-1 {
				flds_str = flds_str[end_pos+1:]
			} else {
				flds_str = ""
			}
		} else {
			pos1 := strings.Index(flds_str, ",")
			pos2 := strings.Index(flds_str, "]")
			if pos1 == -1 && pos2 == -1 {
				val, _ := strconv.Atoi(flds_str)
				ni = append(ni, &NestedInteger{&val, nil})
				flds_str = ""
			} else if pos2 >= 0 {
				var pos int
				if pos2 < pos1 || pos1 == -1 {
					pos = pos2
				} else {
					pos = pos1
				}
				val, _ := strconv.Atoi(flds_str[:pos])
				ni = append(ni, &NestedInteger{&val, nil})
				flds_str = flds_str[pos+1:]
			} else if pos1 >= 0 {
				val, _ := strconv.Atoi(flds_str[:pos1])
				ni = append(ni, &NestedInteger{&val, nil})
				flds_str = flds_str[pos1+1:]
			} else {
				fmt.Printf("createNestedInteger() Error.\n")
			}
		}
	}
	return ni
}

func createNestedInteger(flds_str string) *NestedInteger {
	var ni NestedInteger
	for len(flds_str) > 0 {
		//	fmt.Printf("flds_str = %s\n", flds_str)
		if flds_str[0] == '[' {
			end_pos, depth := 1, 1
			for depth > 0 && end_pos < len(flds_str) {
				if flds_str[end_pos] == '[' {
					depth++
				} else if flds_str[end_pos] == ']' {
					depth--
				}
				end_pos++
			}
			substr := flds_str[1 : end_pos-1]
			ni.Add(*createNestedInteger(substr))
			if end_pos < len(flds_str)-1 {
				flds_str = flds_str[end_pos+1:]
			} else {
				flds_str = ""
			}
		} else {
			pos1 := strings.Index(flds_str, ",")
			pos2 := strings.Index(flds_str, "]")
			if pos1 == -1 && pos2 == -1 {
				val, _ := strconv.Atoi(flds_str)
				ni.Add(NestedInteger{&val, nil})
				flds_str = ""
			} else if pos2 >= 0 {
				var pos int
				if pos2 < pos1 || pos1 == -1 {
					pos = pos2
				} else {
					pos = pos1
				}
				val, _ := strconv.Atoi(flds_str[:pos])
				ni.Add(NestedInteger{&val, nil})
				flds_str = flds_str[pos+1:]

			} else if pos1 >= 0 {
				val, _ := strconv.Atoi(flds_str[:pos1])
				ni.Add(NestedInteger{&val, nil})
				flds_str = flds_str[pos1+1:]
			} else {
				fmt.Printf("createNestedInteger() Error.\n")
			}
		}
	}
	return &ni
}

func listNestedIntegerToString(flds []*NestedInteger) string {
	resultStr := ""
	for _, ni := range flds {
		resultStr += (*ni).ToString()
	}
	return resultStr
}

func getNestedIterator(listNestedInteger []*NestedInteger) []int {
	var result []int
	ni := Constructor(listNestedInteger)
	for ni.HasNext() {
		result = append(result, ni.Next())
	}
	return result
}

File: Problems/0300_0399/0341_Flatten_Nested_List_Iterator/Project_Java/OperateNestedInteger.java
import java.util.*;

public class OperateNestedInteger {
    public List<NestedInteger> createNestedInteger(String flds_str) {
        List<NestedInteger> ni = new ArrayList<>();
        while (flds_str.length() > 0) {
        //  System.out.println("flds_str = " + flds_str);
            if (flds_str.charAt(0) == '[') {
                int end_pos = 1, depth = 1;
                while (depth > 0 && end_pos < flds_str.length()) {
                    if (flds_str.charAt(end_pos) == '[') {
                        depth++;
                    } else if (flds_str.charAt(end_pos) == ']') {
                        depth--;
                    }
                    end_pos++;
                }
                String substr = flds_str.substring(1, end_pos - 1);
                ni.add(createListNestedInteger(substr));
                if (end_pos < flds_str.length() - 1) {
                    flds_str = flds_str.substring(end_pos + 1);
                } else {
                    flds_str = "";
                }
            } else {
                int pos1 = flds_str.indexOf(',');
                int pos2 = flds_str.indexOf(']');
                if (pos1 == -1 && pos2 == -1) {
                     ni.add(new NestedInteger(Integer.parseInt(flds_str)));
                    flds_str = "";
                } else if (pos2 >= 0) {
                    if (pos2 < pos1 || pos1 == -1) {
                        int val = Integer.parseInt(flds_str.substring(0, pos2)); 
                        ni.add(new NestedInteger(val));
                        flds_str = flds_str.substring(pos2 + 1);
                    } else {
                        int val = Integer.parseInt(flds_str.substring(0, pos1)); 
                        ni.add(new NestedInteger(val));
                        flds_str = flds_str.substring(pos1 + 1);
                    }
                } else if (pos1 >= 0) {
                    int val = Integer.parseInt(flds_str.substring(0, pos1)); 
                    ni.add(new NestedInteger(val));
                    flds_str = flds_str.substring(pos1 + 1);
                } else {
                    System.out.println("createNestedInteger() Error.");
                }
            }
        }
        return ni;
    }    

    public NestedInteger createListNestedInteger(String flds_str) {
        NestedInteger ni = new NestedInteger();
        while (flds_str.length() > 0) {
        //  System.out.println("flds_str = " + flds_str);
            while (flds_str.charAt(0) == ',') {
                flds_str = flds_str.substring(1);
            }
            if (flds_str.charAt(0) == '[') {
                int end_pos = 1, depth = 1;
                while (depth > 0 && end_pos < flds_str.length()) {
                    if (flds_str.charAt(end_pos) == '[') {
                        depth++;
                    } else if (flds_str.charAt(end_pos) == ']') {
                        depth--;
                    }
                    end_pos++;
                }
                String substr = flds_str.substring(1, end_pos - 1);
                ni.add(createListNestedInteger(substr));
                if (end_pos < flds_str.length() - 1) {
                    flds_str = flds_str.substring(end_pos + 1);
                } else {
                    flds_str = "";
                }
            } else {
                int pos1 = flds_str.indexOf(',');
                int pos2 = flds_str.indexOf(']');
                if (pos1 == -1 && pos2 == -1) {
                     ni.add(new NestedInteger(Integer.parseInt(flds_str)));
                    flds_str = "";
                } else if (pos2 >= 0) {
                    if (pos2 < pos1 || pos1 == -1) {
                        int val = Integer.parseInt(flds_str.substring(0, pos2)); 
                        ni.add(new NestedInteger(val));
                        flds_str = flds_str.substring(pos2 + 1);
                    } else {
                        int val = Integer.parseInt(flds_str.substring(0, pos1)); 
                        ni.add(new NestedInteger(val));
                        flds_str = flds_str.substring(pos1 + 1);
                    }
                } else if (pos1 >= 0) {
                    int val = Integer.parseInt(flds_str.substring(0, pos1)); 
                    ni.add(new NestedInteger(val));
                    flds_str = flds_str.substring(pos1 + 1);
                } else {
                    System.out.println("createNestedInteger() Error.");
                }
            }
        }
        return ni;
    }

    public String listNestedIntegerToString(List<NestedInteger> flds) {
        StringBuilder sb = new StringBuilder();
        for (NestedInteger ni : flds) {
            sb.append(ni.toString());
        }
        return sb.toString();
    }

    public List<Integer> getNestedIterator(List<NestedInteger> listNestedInteger) {
        List<Integer> result = new ArrayList<>();
        NestedIterator ni = new NestedIterator(listNestedInteger);
        while (ni.hasNext()) {
            result.add(ni.next());
        }
        return result;
    }
}
